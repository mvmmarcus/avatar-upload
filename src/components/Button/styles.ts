import styled, { css, DefaultTheme } from "styled-components";

import { ButtonPros } from ".";

type WrapperProps = ButtonPros;

const wrapperModifiers = {
  link: (theme: DefaultTheme) => css`
    background: transparent;
    text-decoration: underline;
    padding: 0;
    font-size: ${theme.font.sizes.medium};
    line-height: ${theme.font.sizes.xxlarge};
    font-weight: ${theme.font.medium};
    color: ${theme.colors.gray07};

    &:hover {
      background-color: transparent;
      color: ${theme.colors.gray05};
    }
  `,
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, variant }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    text-decoration: none;
    color: #fff;
    border: none;
    transition: background 0.3s ease-in-out;
    height: ${theme.spacings.medium};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.medium};
    background-color: ${theme.colors.gray07};
    border: ${theme.border.radius};
    padding: 0.4rem ${theme.spacings.medium};
    border-radius: ${theme.font.sizes.medium};

    &:hover {
      background-color: ${theme.colors.gray05};
    }

    ${variant === "link" && wrapperModifiers.link(theme)}
  `}
`;
