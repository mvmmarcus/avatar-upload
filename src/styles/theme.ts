export default {
  border: {
    radius: "0.4rem",
  },
  font: {
    family: "Inter, sans-serif",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xsmall: "1.2rem",
      small: "1.4rem",
      medium: "1.6rem",
      large: "1.8rem",
      xlarge: "2.0rem",
      xxlarge: "2.8rem",
    },
  },
  colors: {
    gray01: "#F2F5F8",
    gray04: "#C3CBD5",
    gray05: "#677489",
    gray06: "#495567",
    gray07: "#3D485F",

    blue01: "#3F80FF",
    blue04: "#B9D1FF",
  },
  spacings: {
    xxsmall: "0.8rem",
    xsmall: "1.6rem",
    small: "2.4rem",
    medium: "3.2rem",
    large: "4.0rem",
    xlarge: "4.8rem",
    xxlarge: "5.6rem",
  },
  transition: {
    default: "0.3s ease-in-out",
    fast: "0.1s ease-in-out",
  },
} as const;
