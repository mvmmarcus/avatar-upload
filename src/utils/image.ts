type CropImageProps = {
  imageUrl: string;
  cropWidth?: number;
  preserveAspectRatio?: boolean;
  previewImageRatio?: number;
};

const addImageProcess = async (
  imageSource: string
): Promise<HTMLImageElement> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = imageSource;
  });
};

export const cropImage = async ({
  imageUrl,
  cropWidth = 300,
  preserveAspectRatio = true,
  previewImageRatio = 1,
}: CropImageProps) => {
  const image = await addImageProcess(imageUrl);

  const ratio = image.height / image.width;

  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");

  if (preserveAspectRatio) {
    canvas.width = image.width < cropWidth ? image.width : cropWidth;
    canvas.height = image.width < cropWidth ? image.height : cropWidth * ratio;

    context?.drawImage(image, 0, 0, canvas.width, canvas.height);
  } else {
    canvas.width = previewImageRatio * image.width;
    canvas.height = canvas.width;

    const cropDistanceX = (image.width - canvas.width) / 2;
    const cropDistanceY = (image.height - canvas.height) / 2;

    context?.drawImage(
      image,
      cropDistanceX,
      cropDistanceY,
      canvas.width,
      canvas.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
  }

  const croppedImageUrl = canvas.toDataURL();

  return croppedImageUrl;
};

export const uploadImage = (file: File): Promise<ProgressEvent<FileReader>> => {
  return new Promise((resolve, reject) => {
    const fileReader = new FileReader();

    fileReader.onload = (event) => resolve(event);
    fileReader.onerror = reject;
    fileReader.readAsDataURL(file);
  });
};
