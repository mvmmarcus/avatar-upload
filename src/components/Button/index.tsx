import React from "react";

import { ButtonHTMLAttributes } from "react";

import { Wrapper } from "./styles";

type ButtonTypes = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  variant?: "link" | "default";
} & ButtonTypes;

const Button = ({ children, ...props }: ButtonProps) => (
  <Wrapper {...props}>{children}</Wrapper>
);

export default Button;
