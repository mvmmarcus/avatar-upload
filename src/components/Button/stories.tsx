import React from "react";

import { Story, Meta } from "@storybook/react/types-6-0";

import Button, { ButtonProps } from ".";

export default {
  title: "Button",
  component: Button,
  argTypes: {
    children: {
      type: "string",
    },
    variant: {
      control: {
        type: "select",
        options: ["link", "default"],
      },
    },
  },
} as Meta<ButtonProps>;

export const Default: Story<ButtonProps> = (args) => <Button {...args} />;

Default.args = {
  children: "Buy now",
  variant: "default",
};

export const Link: Story<ButtonProps> = (args) => <Button {...args} />;

Link.args = {
  children: "Try again",
  variant: "link",
};
