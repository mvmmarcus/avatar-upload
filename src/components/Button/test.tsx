import React from "react";

import { screen } from "@testing-library/react";

import Button from ".";
import theme from "styles/theme";
import { renderWithTheme } from "utils/test-utils";

describe("<Button />", () => {
  it("should render default button", () => {
    const { container } = renderWithTheme(<Button>Save</Button>);

    expect(screen.getByRole("button", { name: /Save/i })).toHaveStyle({
      "background-color": theme.colors.gray07,
    });

    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render link button", () => {
    const { container } = renderWithTheme(<Button variant="link">Save</Button>);

    expect(screen.getByRole("button", { name: /Save/i })).toHaveStyle({
      background: "transparent",
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
