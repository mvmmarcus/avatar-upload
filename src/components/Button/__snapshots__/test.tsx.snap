// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render default button 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
}

.c0:hover {
  background-color: #677489;
}

<button
  class="c0"
>
  Save
</button>
`;

exports[`<Button /> should render link button 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
  background: transparent;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  padding: 0;
  font-size: 1.6rem;
  line-height: 2.8rem;
  font-weight: 500;
  color: #3D485F;
}

.c0:hover {
  background-color: #677489;
}

.c0:hover {
  background-color: transparent;
  color: #677489;
}

<button
  class="c0"
>
  Save
</button>
`;
